hazelcast/src/main/java/com/hazelcast/topic/impl/TopicProxy.java

```diff
@@ -47,6 +56,13 @@ public void publish(@Nonnull E message) {
         publishInternal(message);
     }
 
+    @Override
+    public InternalCompletableFuture<E> publishAsync(@Nonnull E message) {
```

Comment:
Another layer of complexity :) In addition to reliable topic vs topic, "topic" can be different if we're talking about "totally-ordered" topic and "non-totally-ordered" topic (determined by `TopicConfig#isGlobalOrderingEnabled`). The "non-ordered" is implemented in `TopicProxy` and `ClientTopicProxy` and here you avoid firing any partition operations since partition operations are ordered. On the other hand, in `TotalOrderedTopicProxy` you can override the method from `TopicProxy` and fire the `PublishAllOperation` (but remember to check for the cluster version first).

Looks like clients don't have non-ordered topics, they send a message task which then sends a publish operation. Basically turning it into an ordered topic always.

This also means that the async member-side non-ordered implementation of `TopicProxy#publishAllAsync` is basically the same as the sync one, since the `TopicProxySupport#publishInternal` doesn't *usually* block. Usually meaning there's some silly rule that every 100000th event should block. This IMHO is a poor backpressure mechanism and I think we should remove it anyway.

Reply:
Ok, here I'm lost... Do you mean, that I should implement the missing methods in the class `TotalOrderedTopicProxy`? And which clients do you think of?

Reply:
When talking about `ITopic` implementations, there are several:
- `TopicProxy` - member side implementation, no ordering guarantees, events are published through `EventService#publishEvent` which means every member can notify every other member
- `TotalOrderedTopicProxy` - member side implementation, an extension of `TopicProxy` which is used when `TopicConfig#isGlobalOrderingEnabled`, doesn't publish events locally through `EventService#publishEvent` but sends partition operations to the same partition so events published on all members are ordered
- `ClientTopicProxy` - client side implementation, invokes operations on the partition owner for the topic name, which means the events are ordered. That member, on the other hand, calls the `EventService#publishEvent` in `PublishOperation#run`
- `ReliableTopicProxy`, `ClientReliableTopicProxy` - use ringbuffer instead of `EventService#publishEvent`, all implementations order messages because they use a single ringbuffer as a backing structure. Since it doesn't use `EventService`, it's not supposed to be lossy

If I remember correctly, what I meant to say is - don't invoke partition operations then in `TopicProxy`, use the `EventService#publishEvent`. It is already async so probably your sync and async implementation in `TopicProxy` will be the same. On the other hand, you should invoke partition operations in `TotalOrderedTopicProxy` since there events published from different members will then be serialized and published through a single partition owner.

If it still doesn't make sense, please tell me, I'll take a closer look and provide some code snippets, I just didn't want to keep you waiting for months :)

Reply:
Something like this:
```java
public class TopicProxy<E> extends TopicProxySupport implements ITopic<E> {

    ...
    protected final SerializationService serializationService;

    public TopicProxy(String name, NodeEngine nodeEngine, TopicService service) {
        super(name, nodeEngine, service);
        this.serializationService = nodeEngine.getSerializationService();
    }

    @Override
    public void publish(@Nonnull E message) {
        checkNotNull(message, NULL_MESSAGE_IS_NOT_ALLOWED);
        publishInternal(message);
    }

    @Override
    public CompletionStage<E> publishAsync(@Nonnull E message) {
        publish(message);
        return InternalCompletableFuture.completedFuture(null);
    }
    ...
}
```


```java
public class TotalOrderedTopicProxy<E> extends TopicProxy<E> {

    ...

    @Override
    public void publish(@Nonnull  E message) {
        checkNotNull(message, NULL_MESSAGE_IS_NOT_ALLOWED);
        Operation operation = new PublishOperation(getName(), toData(message))
                .setPartitionId(partitionId);
        InternalCompletableFuture f = invokeOnPartition(operation);
        f.joinInternal();
    }

    @Override
    public CompletionStage<E> publishAsync(@Nonnull E message) {
        checkNotNull(message, NULL_MESSAGE_IS_NOT_ALLOWED);
        Operation op = new PublishOperation(getName(), toData(message));
        return newDelegatingFuture(serializationService, putAsyncInternal(op));
    }
}
```